static/src/javascripts/projects/common/modules/component.js

```diff
@@ -0,0 +1,361 @@
+                });
+        };
+
+        const setAutoUpdate = () => {
+            this.t = setTimeout(() => update(), this.updateEvery * 1000);
+        };
+
+        if (this.autoupdated) {
+            setAutoUpdate();
+        }
+    }
+
+    /**
+     * This is user to edit this.elem before it's rendered
+     * This will help with the rendering performance that
+     * we would lose if rendered then manipulated
+     */
+    prerender(): void {}
+
+    /**
+     * Once the render / decorate methods have been called
+     * This is where you could do your event binding
+     * This function is made to be overridden
+     */
+    ready(): void {}
+
+    /**
+     * Once the render / decorate methods have been called
+     * This is where you could do your error event binding
+     * This function is made to be overridden
+     */
+    error(type?: string, message?: string): void {}
+
+    /**
+     * This is called whenever a fetch occurs. This includes
+     * explicit fetch calls and autoupdate.
+     */
+    fetched(resp?: Object): void {}
+
+    autoupdate(elem: HTMLElement): void {
+        const oldElem = this.elem;
+        this.elem = elem;
+
+        this._prerender();
+        bonzo(oldElem).replaceWith(this.elem);
+    }
+
+    /**
+     * Once we're done with it, remove event bindings etc
+     */
+    dispose(): void {}
+
+    on(
+        eventType: string,
+        elem: string | ((event: Event) => void),
+        handler?: (event: Event) => void
+    ): Component {
+        if (typeof elem === 'function') {
+            handler = elem;
+            bean.on(this.elem, eventType, handler.bind(this));
+        } else {
+            elem = !elem.length ? [elem] : elem;
+            bean.on(this.elem, eventType, elem, handler.bind(this));
+        }
+
+        return this;
+    }
+
+    emit(eventName: string, args?: Object): void {
+        bean.fire(this.elem, eventName, args);
+    }
+
+    getElem(elemName: string): HTMLElement {
+        if (this.elems && this.elems[elemName]) {
+            return this.elems[elemName];
+        }
+
+        const elem = qwery(this.getClass(elemName), this.elem)[0];
+
+        if (this.elems) {
+            this.elems[elemName] = elem;
+        }
+
+        return elem;
+    }
+
+    getClass(elemName: string, sansDot: boolean = false): string {
+        let className;
+
+        if (this.useBem && this.componentClass) {
+            className = `${this.componentClass}__${elemName}`;
+        } else {
+            className = (this.classes && this.classes[elemName]) || '';
+        }
+
+        return (sansDot ? '' : '.') + className;
+    }
+
+    setState(state: string, elemName: ?string) {
+        const elem = elemName ? this.getElem(elemName) : this.elem;
+        const $elem = bonzo(elem);
+
+        if (this.componentClass) {
+            $elem.addClass(
+                `${this.componentClass +
+                    (elemName ? `__${elemName}` : '')}--${state}`
+            );
```

Comment:
What do you think about extracting this little bit of classname logic into a function?